<?xml version="1.0" encoding="UTF-8"?>
<!-- Top-level launch file for Design nr.3, -->

<launch>

    <!-- Run the map server to load a pre-made map -->
    <arg name="map_file" default="$(find final_project)/maps/final_project_world.yaml"/>
    <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" />

    <!--- Start AMCL Localization -->
    <arg name="odom_frame_id"   default="/odom"/>
    <arg name="base_frame_id"   default="/base_footprint"/>
    <arg name="global_frame_id" default="map"/>
    <arg name="use_map_topic"   default="false"/>
    <arg name="scan_topic"      default="/kobuki/laser/scan"/>

    <arg name="initial_pose_x"  default="0.0"/> <!-- in meters -->
    <arg name="initial_pose_y"  default="8.0"/> <!-- in meters -->
    <arg name="initial_pose_a"  default="0.0"/> <!-- in radians -->

    <arg name="initial_cov_xx"  default="0.25"/>
    <arg name="initial_cov_yy"  default="0.25"/>
    <arg name="initial_cov_aa"  default="0.0685"/> <!-- default: (π/12)*(π/12) radian. 2*PI = 6.28 radians -->

    <arg name="min_particles"   default="100"/>
    <arg name="max_particles"   default="2000"/>

    <arg name="laser_max_beams" default="30"/>
    <arg name="laser_model_type" default="likelihood_field"/>
    <arg name="laser_likelihood_max_dist" default="2.0"/>

    <arg name="recovery_alpha_slow" default="0.0"/>
    <arg name="recovery_alpha_fast" default="0.0"/>
    <arg name="resample_interval" default="2"/>
    <arg name="kld_err" default="0.01"/>  
    <arg name="kld_z" default="0.99"/>

    <node pkg="amcl" type="amcl" name="amcl">
        <param name="use_map_topic"             value="$(arg use_map_topic)"/>
        <!-- Publish scans from best pose at a max of 10 Hz -->
        <param name="odom_model_type"           value="diff-corrected"/>
        <param name="odom_alpha5"               value="0.1"/>
        <param name="gui_publish_rate"          value="10.0"/>
        <param name="laser_max_beams"           value="$(arg laser_max_beams)"/>
        <param name="laser_max_range"           value="12.0"/>
        <param name="min_particles"             value="$(arg min_particles)"/>
        <param name="max_particles"             value="$(arg max_particles)"/>
        <param name="kld_err"                   value="$(arg kld_err)"/>
        <param name="kld_z"                     value="$(arg kld_z)"/>
        <param name="odom_alpha1"               value="0.05"/>
        <param name="odom_alpha2"               value="0.05"/>
        <!-- translation std dev, m -->
        <param name="odom_alpha3"               value="0.10"/>
        <param name="odom_alpha4"               value="0.05"/>
        <param name="laser_z_hit"               value="0.95"/>
        <param name="laser_z_short"             value="0.1"/>
        <param name="laser_z_max"               value="0.05"/>
        <param name="laser_z_rand"              value="0.05"/>
        <param name="laser_sigma_hit"           value="0.2"/> <!-- param -->
        <param name="laser_lambda_short"        value="0.1"/>
        <param name="laser_model_type"          value="$(arg laser_model_type)"/>
        <param name="laser_likelihood_max_dist" value="$(arg laser_likelihood_max_dist)"/>
        <param name="update_min_d"              value="0.25"/>
        <param name="update_min_a"              value="0.2"/>
        <param name="odom_frame_id"             value="$(arg odom_frame_id)"/>
        <param name="base_frame_id"             value="$(arg base_frame_id)"/>
        <param name="global_frame_id"           value="$(arg global_frame_id)"/>
        <!-- larger resample_interval values lead to slower particle convergence -->
        <param name="resample_interval"         value="$(arg resample_interval)"/>
        <!-- Increase tolerance because the computer can get quite busy -->
        <param name="transform_tolerance"       value="1.0"/>
        <param name="recovery_alpha_slow"       value="$(arg recovery_alpha_slow)"/>
        <param name="recovery_alpha_fast"       value="$(arg recovery_alpha_fast)"/>
        <param name="initial_pose_x"            value="$(arg initial_pose_x)"/>
        <param name="initial_pose_y"            value="$(arg initial_pose_y)"/>
        <param name="initial_pose_a"            value="$(arg initial_pose_a)"/>
        <param name="initial_cov_xx"            value="$(arg initial_cov_xx)"/>
        <param name="initial_cov_yy"            value="$(arg initial_cov_yy)"/>
        <param name="initial_cov_aa"            value="$(arg initial_cov_aa)"/>
        <remap from="scan"                      to="$(arg scan_topic)"/>
        <remap from="initialpose"               to="/initialpose"/>
        <remap from="amcl_pose"                 to="/amcl_pose"/>
        <remap from="particlecloud"             to="/particlecloud"/>

    </node>

    <!-- EKF Local -->
    <node pkg="robot_localization" type="ekf_localization_node" name="local_ekf" clear_params="true"> 
      <param name="frequency" value="50"/>  
      <param name="sensor_timeout" value="0.1"/>  
      <param name="two_d_mode" value="true"/>
      <!-- Outputs map->odom transform (if enabled)-->
      <param name="publish_tf" value="false"/>
      <param name="odom_frame" value="odom"/>
      <param name="map_frame" value="map"/>
      <param name="base_link_frame" value="base_footprint"/>
      <!-- local EKF uses odom as world frame -->
      <param name="world_frame" value="odom"/>
      <rosparam param="initial_state">[0.0,  8.0,  0.0,
                                 0.0,  0.0,  0.0,
                                 0.0,  0.0,  0.0,
                                 0.0,  0.0,  0.0,
                                 0.0,  0.0,  0.0]</rosparam>

      <!-- Inputs --> 
      <param name="imu0" value="/mobile_base/sensors/imu_data_raw"/> 
      <rosparam param="imu0_config">[false, false, false, 
                                     false, false, true,
                                     false, false, false,
                                     false, false, true,
                                     true, false, false]</rosparam>
      <param name="imu0_differential" value="false"/>
      <param name="odom0" value="/odom"/>
      <rosparam param="odom0_config">[false, false, false,
                                      false, false, true,
                                      true, true, false,
                                      false, false, true,
                                      false, false, false]</rosparam>
      <param name="odom0_differential" value="false"/>

      <rosparam param="process_noise_covariance">[0.05, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                              0,    0.05, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                              0,    0,    0.06, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                              0,    0,    0,    0.03, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                              0,    0,    0,    0,    0.03, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                              0,    0,    0,    0,    0,    0.06, 0,     0,     0,    0,    0,    0,    0,    0,    0,
                                              0,    0,    0,    0,    0,    0,    0.025, 0,     0,    0,    0,    0,    0,    0,    0,
                                              0,    0,    0,    0,    0,    0,    0,     0.025, 0,    0,    0,    0,    0,    0,    0,
                                              0,    0,    0,    0,    0,    0,    0,     0,     0.04, 0,    0,    0,    0,    0,    0,
                                              0,    0,    0,    0,    0,    0,    0,     0,     0,    0.01, 0,    0,    0,    0,    0,
                                              0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0.01, 0,    0,    0,    0,
                                              0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0.02, 0,    0,    0,
                                              0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0.01, 0,    0,
                                              0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0.01, 0,
                                              0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0.015]</rosparam>

    <rosparam param="initial_estimate_covariance"> [1e-9, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                      0,    1e-9, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                      0,    0,    1e-9, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                      0,    0,    0,    1e-9, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                      0,    0,    0,    0,    1e-9, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                      0,    0,    0,    0,    0,    1e-9, 0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                      0,    0,    0,    0,    0,    0,    1e-9, 0,    0,    0,     0,     0,     0,    0,    0,
                                                      0,    0,    0,    0,    0,    0,    0,    1e-9, 0,    0,     0,     0,     0,    0,    0,
                                                      0,    0,    0,    0,    0,    0,    0,    0,    1e-9, 0,     0,     0,     0,    0,    0,
                                                      0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-9,  0,     0,     0,    0,    0,
                                                      0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-9,  0,     0,    0,    0,
                                                      0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-9,  0,    0,    0,
                                                      0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-9, 0,    0,
                                                      0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-9, 0,
                                                      0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-9]</rosparam>
      <!-- Outputs Default topics are: odometry/filtered, accel/filtered (if enabled) -->
      <remap from="odometry/filtered" to="odometry/filtered" />

    </node>

  <!-- Determine whether to run Rviz or not & set Rviz config file-->
  <arg name="use_rviz" default="true" />
  <arg name="rvizconfig" default="$(find final_project)/rviz/design3.rviz"/>
  <!-- Launching Rviz -->
  <node if="$(arg use_rviz)" name="rviz" pkg="rviz" type="rviz" respawn="false" args="-d $(arg rvizconfig)"/>




</launch>
